src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs

```diff
@@ -116,10 +117,143 @@ private void Visit(IAssemblySymbol assembly)
+                        SyntaxFactory.VariableDeclarator(
+                            SyntaxFactory.Identifier(fieldName))))
+            ).WithModifiers(SyntaxFactory.TokenList(new []{SyntaxFactory.Token(SyntaxKind.PrivateKeyword)})
+            ).WithAttributeLists(attrs);
+        }
+
+        // SynthesizeDummyFields yields private fields for the namedType, because they can be part of the API contract.
+        // - A struct containing a field that is a reference type cannot be used as a reference.
+        // - A struct containing nonempty fields needs to be fully initialized. (See "definite assignment" rules)
+        //   - "non-empty" means either unmanaged types like ints and enums, or reference types that are not the root.
+        // - A struct containing generic fields cannot have struct layout cycles.
+        private static IEnumerable<SyntaxNode> SynthesizeDummyFields(ISymbolFilter symbolFilter, INamedTypeSymbol namedType)
+        {
+            // Collect all excluded fields
+            IEnumerable<IFieldSymbol> excludedFields = namedType.GetMembers()
+                .Where(member => !symbolFilter.Include(member) && member is IFieldSymbol)
+                .Select(m => (IFieldSymbol)m);
+            
+            if (excludedFields.Any())
+            {
+                // Collect generic excluded fields
+                IEnumerable<IFieldSymbol> genericTypedFields = excludedFields.Where(f => {
+                    if (f.Type is INamedTypeSymbol ty) {
+                        return ty.IsGenericType;
+                    }
+                    return f.Type is ITypeParameterSymbol;
+                });
+
+                // Add a dummy field for each generic excluded field
+                foreach(IFieldSymbol genericField in genericTypedFields)
+                {
+                    yield return CreateDummyField(
+                        genericField.Type.ToDisplayString(),
+                        genericField.Name,
+                        FromAttributeData(genericField.GetAttributes()));
+                }
+
+                bool hasRefPrivateField = excludedFields.Any(f => IsOrContainsReferenceType(f.Type));
+
+                // If any field's type is transitively a reference type.
+                if (hasRefPrivateField)
+                {
+                    // add reference type dummy field
+                    yield return CreateDummyField("object", "_dummy", new());
+
+                    // add int field
+                    yield return CreateDummyField("int", "_dummyPrimitive", new());
+                }
+                else
+                {
+                    bool hasNonEmptyStructPrivateField = excludedFields.Any(f => IsOrContainsNonEmptyStruct(f.Type));
+                    
+                    // Otherwise, if the field transitively contains a field whose type is non-empty.
+                    if (hasNonEmptyStructPrivateField)
```

Comment:
The bool value is only used here. Inline it:

```suggestion
                    // Otherwise, if the field transitively contains a field whose type is non-empty.
                    if (excludedFields.Any(excludedField => IsOrContainsNonEmptyStruct(excludedField.Type)))
```